File: 3-Structural Patterns/5-Decorator Pattern/Ice Cream Example/Implementation with Factories/Java/runner/DecoratorPatternTest.java
package runner;

import ice_cream.IceCream;
import ice_cream_factory.IceCreamWithNutsFactory;
import ice_cream_factory.IceCreamWithSprinklesFactory;
import ice_cream_factory.VanillaIceCreamFactory;

/**
 * Application that actually uses Decorator Pattern.
 *
 * @author Ziang Lu
 */
public class DecoratorPatternTest {

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        IceCream myIceCream = new IceCreamWithNutsFactory(
                new IceCreamWithSprinklesFactory(VanillaIceCreamFactory.getInstance())).makeIceCream();
        System.out.println(myIceCream);

        /*
         * Output:
         * Creating a Vanilla Ice Cream
         * Adding Sprinkles
         * Adding Nuts
         * VanillaIceCream with toppings [Sprinkles, Nuts]
         */
    }

}

File: 3-Structural Patterns/5-Decorator Pattern/Ice Cream Example/Implementation with Factories/Python/decorator_pattern_test.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Application that actually uses Decorator Pattern.
"""

__author__ = 'Ziang Lu'

from ice_cream_factory import (
    IceCreamWithNutsFactory, IceCreamWithSprinklesFactory,
    VanillaIceCreamFactory
)


def main():
    my_ice_cream = IceCreamWithNutsFactory(
        IceCreamWithSprinklesFactory(VanillaIceCreamFactory.get_instance())
    ).make_ice_cream()
    print(my_ice_cream)


if __name__ == '__main__':
    main()

# Output:
# Creating a Vanilla Ice Cream
# Adding Sprinkles
# Adding Nuts
# VanillaIceCream with toppings ['Sprinkles', 'Nuts']
